### PR TITLE
Feature/images by tag

### LINKED ISSUE #7: Images by Tag
#### As a user I want to view all images associated with a tag.

__Story__:

Now that we have added the ability for images to be associated with tags, let
us provide the ability for users to filter images by a given tag. Nana, who
loves fast planes, will be delighted to see only images labeled with SR-71.

__Acceptance criteria__:
- [x] When I click on a tag, I see a filtered list of only the images that have
  the tag I clicked on.
- [x] The tag name, not its ID, is used to perform the filtering.

__Discussion topics__:
- [ ] Where in the code should we implement this functionality?
- [ ] What happens if a user tries to manually filter by a nonexistent tag?

__Dependencies__:
- Image Tags

### PR DESCRIPTION
**Pull request by jesse-smidt**
Closes #7 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -15,8 +15,8 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
     follow_redirect!
     assert_response :success
     assert_select 'a', URL_PATH
-    assert_select 'span.badge:nth-of-type(1)', text: 'nice'
-    assert_select 'span.badge:nth-of-type(2)', text: 'fun'
+    assert_select 'a.badge:nth-of-type(1)', text: 'nice'
```
Comment (stevenboyd) [inline test/controllers/images_controller_test.rb]: I would also test that the link is correct

Comment (stevenboyd) [reply test/controllers/images_controller_test.rb]: Also add a test to the index controller

File: app/views/application/home.html.erb [hunk 2]
```diff
@@ -30,3 +32,9 @@
     </div>
   <% end %>
 </div>
+<% if @images.empty? %>
```
Comment (stevenboyd) [inline app/views/application/home.html.erb]: You should add a test for this case.

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,6 +1,7 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   def home
-    @images = Image.order(created_at: :desc)
+    images = Image.tagged_with(params[:tag]) if params[:tag]
```
Comment (stevenboyd) [inline app/controllers/application_controller.rb]: I would write this a different way. I think a cleaner way is to do
```ruby
images = Images.where
images = image.tagged_with(params[:tag]) if params[:tag]
@images = images.order(created_at: :desc)
```

This relies on the fact that all these methods return an ActiveRecord::Relation object. You can read a brief into here: https://guides.rubyonrails.org/active_record_querying.html#understanding-the-method-chaining

Also, I would add some test cases to the home controllers test to cover tag filtering.

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (jesse-smidt) [reply test/controllers/images_controller_test.rb]: Added checks

File: app/views/application/home.html.erb (hunk 2 above)
Comment (jesse-smidt) [reply app/views/application/home.html.erb]: Added test

File: app/controllers/application_controller.rb (hunk 3 above)
Comment (jesse-smidt) [reply app/controllers/application_controller.rb]: Updated

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,7 +1,3 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
-  def home
```
Comment (stevenboyd) [inline app/controllers/application_controller.rb]: Good refactor!

File: test/controllers/application_controller_test.rb [hunk 5]
```diff
@@ -2,11 +2,20 @@
...
 class ApplicationControllerTest < ActionDispatch::IntegrationTest
   test 'should get home with image with tags' do
-    get '/'
+    get root_url
 
     assert_response :success
     assert_select 'h1', text: 'Image Sharer'
     assert_select '.card img[src]'
-    assert_select '.card .badge', text: 'fun'
+    assert_select '.card .badge[href=?]', root_url + '?tag=fun', text: 'fun'
```
Comment (stevenboyd) [inline test/controllers/application_controller_test.rb]: For these cases, I would do `root_url(tag: 'fun')` instead of building the URL yourself.

Comment (jesse-smidt) [reply test/controllers/application_controller_test.rb]: Awesome - these route helpers are pretty flexible...

**Pull request merged**